src/psyclone/psyir/symbols/datatypes.py

```diff
@@ -335,8 +380,33 @@ def __eq__(self, other):
         if not super().__eq__(other):
             return False
 
-        return (self.precision == other.precision and
-                self.intrinsic == other.intrinsic)
+        if isinstance(other.precision, Symbol) and isinstance(self.precision,
+                                                              Symbol):
+            # If the precision in both types is given by a Symbol then we just
+            # compare the names of those symbols.
+            precision_match = other.precision.name == self.precision.name
+        else:
+            precision_match = self.precision == other.precision
+        return precision_match and self.intrinsic == other.intrinsic
```

Comment:
One instance of the problem is in the kernel extraction tests:
```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> captured stdout >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
ARPDBG: Added i_def 140716787661664 to table of InvokeSchedule:
End InvokeSchedule

ARPDBG: Added i_def 140716783407840 to table of Routine[name:'testkern_coord_w0_2_code']:
End Routine
ARPDBG: Added i_def 140716783415616 to table of None
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
(Pdb) id(datatype.precision)
140716787661664
(Pdb) id(sym.datatype.precision)
140716783415616
```

Reply:
So we have two different `LFRicIntegerScalarDataType` instances, each containing a different 'i_def' symbol as its precision. Those two should be the *same* symbol.

Reply:
```
PSyclone/src/psyclone/domain/lfric/lfric_extract_driver_creator.py:932: in create
    child.lower_to_language_level()
PSyclone/src/psyclone/domain/lfric/lfric_loop.py:140: in lower_to_language_level
    start = self.start_expr.copy()
PSyclone/src/psyclone/domain/lfric/lfric_loop.py:780: in start_expr
    lbound = sym_table.find_or_create_integer_symbol(root_name,
```
It's the "new" `datatype` created in the last routine that has the reference to `i_def` in the original InvokeSchedule. However, at this point, we're dealing with a copy of that schedule (as we're in the process of making a new driver).

Reply:
I've create #2659 and added a TODO.

Reply:
All in all, that was another good spot so thanks. I've tightened up what I have so that it also checks that the interface of the two precision symbols matches.